lib/improver/argparser.py

```diff
@@ -28,19 +28,185 @@
 # CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
 # ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
 # POSSIBILITY OF SUCH DAMAGE.
-"""Common option utilities for Improver CLIs."""
+"""Common option utilities for improver CLIs."""
 
-import argparse
+from argparse import ArgumentParser
 
 
-class ArgParser(argparse.ArgumentParser):
-    """Argument parser for improver CLIs"""
+class ArgParser(ArgumentParser):
+    """Argument parser for improver CLIs.
+    (Subclass of argparse.ArgumentParser)"""
+
+    # CENTRALIZED_ARGUMENTS, is a dictionary where, for each element:
+    #   the key is a string representing the argument name,
+    #   the value is a list containing 2 elements:
+    #       1. a list of strings containing the different flags
+    #       2. a dictionary containing all of the kwargs which are passed
+    #          to the ArgumentParser.add_argument() method
+
+    # All CLIs should select something from this dictionary:
+    # NB: --help included for free with ArgumentParser
+    CENTRALIZED_ARGUMENTS = {
+        'input_file': (
+            ['input_filepath'],
+            {}),
+        'output_file': (
+            ['output_filepath'],
+            {}),
+    }
+
+    # *All* CLIs will use the options here (no option to disable them):
+    COMPULSORY_ARGUMENTS = {
+        # TODO: Implement (some of?) these:
+        #        'debug': (
+        #            ['--debug'],
+        #            {'action': 'store_true'}),
+        #        'dry_run': (
+        #            ['--dry-run'],
+        #            {'action': 'store_true'}),
+        #        'profile' : (
+        #            ['--profile'],
+        #            {'action': 'store_true'}),
+        #        'verbose': (
+        #            ['--verbose', '-v'],
+        #            {'action': 'count'}),
+        #        'version': (
+        #            ['--version'],
+        #            {'action': 'version',
+        #             'version': improver.version_message() }),
+    }
+
+    # We can override including these, but options common to everything should
+    # be in a list here:
+    # DEFAULT_CENTRALIZED_ARGUMENTS = ('input_file', 'output_file')
+    DEFAULT_CENTRALIZED_ARGUMENTS = ()
```

Comment:
CENTRALISED_ARGUMENTS and DEFAULT_CENTRALISED_ARGUMENTS at the moment have different types.  Could we call this something different to make that explicit?  Eg DEFAULT_CENTRALISED_ARG_NAMES.

Reply:
Sure. Will do.

Reply:
So these are the names of the centralised options we want to use for all CLIs, but each CLI can choose to modify these. Hopefully this will be made clear in the centralised docstring Caroline has asked for.

Reply:
Almost.

A CLI will select from the collection of centralized options, but they are taken "as-is" (cannot be modified). The compulsory options (currently an empty dict) are CLI options which are globally included. The "specific_arguments" are the ones which are specific to this particular CLI, and not used frequently enough to warrant being added to the centralized dictionary.

The idea is that common options are added to the centralized dictionary once they're used more than a certain number of times, to reduce duplication and more closely align the CLIs, and the compulsory options are things like profiling and verbosity options, which we'll want for everything.